### ISSUE TITLE
A function that updates the plot limits of a plot in place without plotting new stuff.

### ISSUE THREAD

**Issue by Datseris** [JuliaPlots/AbstractPlotting.jl#62]
```
scene = scatter(rand(10), rand(10))
limits = FRect(0.1,0.1,0.1,0.1)
AbstractPlotting.update_limits!(scene, limits)
```
gives
![image](https://user-images.githubusercontent.com/19669089/50214806-74fe5400-0381-11e9-97ef-46d4a8f59ee2.png)

I really need the above for **DynamicalSystems.jl** :)

**Comment by SimonDanisch** [comment]
What is the exact thing you need here? cropping the scatter points outside limits? Or zooming to the new axis?

**Comment by Datseris** [comment]
Cropping is not really important. Zooming into the new axis is tje important part!

But of course the scaling will also change , because I want to be able to zoom into rectangles , nnot only perfect squares.

This means that this function must also update the axis ticks , exactly like the one from PyPlot

**Comment by SimonDanisch** [comment]
Ok, there is some issue with scaled scenes in `update_cam!` - really odd that this hasn't been a problem in previous examples... I'll look into it

**Comment by SimonDanisch** [comment]
Sorry for the slow response. As a quick workaround you can use:
```Julia
using Makie
using AbstractPlotting: modelmatrix
scene = scatter(rand(10), rand(10))
limits = FRect(0.1,0.1,0.1,0.1)
AbstractPlotting.update_limits!(scene, limits)
# or alternative focus on bounds of axis 
# (instead of limits use boundingbox(scene[Axis]))
AbstractPlotting.update_cam!(scene, (modelmatrix(scene)) * limits)
```

**Comment by Datseris** [comment]
Hm, unfortunately this doesn't quite work...
Here is an example:
```
using Makie
using AbstractPlotting: modelmatrix
scene = scatter([0, 0.1, 1], [0, 0.1, 1])
```
![image](https://user-images.githubusercontent.com/19669089/50825789-33d2e180-133a-11e9-839e-2be196f7d836.png)

After
```
limits = FRect(0.1,0.1,0.1,0.1)
AbstractPlotting.update_limits!(scene, limits)
```
![image](https://user-images.githubusercontent.com/19669089/50825818-4816de80-133a-11e9-9855-f322cdb2c87f.png)
(we already know this doesn't work).

After you last suggestion:
```
AbstractPlotting.update_cam!(scene, (modelmatrix(scene)) * limits)
```
![image](https://user-images.githubusercontent.com/19669089/50825859-6977ca80-133a-11e9-8884-620ed16fd1a2.png)

As you can see the axis are completely hidden. 

**After zooming out using the mousewheel**

![image](https://user-images.githubusercontent.com/19669089/50825887-814f4e80-133a-11e9-854b-4f6b16f1a61a.png)

So this needs me to physically touch the mouse and zoom out, thus I think there is still a tiny bit left of a fix to be done.

(Totally off-topic, but I think that from this example it becomes very clear that we need dimensionless plotting units; this ball is humongous)

**Comment by SimonDanisch** [comment]
Please read my suggestions and don't omit stuff from my code samples!
```
# or alternative focus on bounds of axis 
# (instead of limits use boundingbox(scene[Axis]))
```

**Comment by Datseris** [comment]
Here is the answer for anyone following this
```
julia>  scene = scatter(rand(100), rand(100)

julia> AbstractPlotting.update_limits!(scene, limits) # command 1

julia> AbstractPlotting.update_cam!(scene, (modelmatrix(scene)) * boundingbox(scene[Axis])) # command 2
```
![image](https://user-images.githubusercontent.com/19669089/50828449-b8752e00-1341-11e9-884f-1cb1250e5df7.png)

I do read your suggestions by the way, and I can suggest you to be clearer with them.

---

Alright, so this is the answer. There are only two things left: there should be a function that does both of these things automatically. The name `update_limits!` for me feels like it should do both of the commands. If you disagree let's just create a new name that runs both the commands. 

The second point is what you can see in the figure I attached. I now realize that this zooming should also crop the data... The entirety of the left axis is hidden and thus the plot is not usable. I know I said that cropping shouldn't really matter, but I didn't think if such cases (which are common).

**Comment by Datseris** [comment]
bump about whether `update_limits` should be updated to do both commands

**Comment by SimonDanisch** [comment]
maybe we should put that in `update!` ...

**Comment by Datseris** [comment]
when you say `update!` do you mean `update_limits!` and thus agree with my proposal, or you mean to create a new function called just `update!` ?

**Comment by Datseris** [comment]
The solution 
```
julia>  scene = scatter(rand(100), rand(100)

julia> AbstractPlotting.update_limits!(scene, limits) # command 1

julia> AbstractPlotting.update_cam!(scene, (modelmatrix(scene)) * boundingbox(scene[Axis])) # command 2
```
that I've posted here does not work for the orbit diagram problem discussed in #65 . I guess it doesn't work if you have multiple subplots... I have the same problem as in the comment: https://github.com/JuliaPlots/AbstractPlotting.jl/pull/65#issuecomment-453732443

selecting a small slice (i.e small x span but large y span) just does not re-adjust the plot to cover all the available figure space.

**Comment by Datseris** [comment]
Simon has given the following code:
```
        AbstractPlotting.update_limits!(scplot, rect)
        AbstractPlotting.scale_scene!(scplot)
        AbstractPlotting.center!(scplot)
        AbstractPlotting.update!(scplot)
```
but it doesn't work 100% of the time. (I don't know why)

**Comment by SimonDanisch** [comment]
Screenshot, or a way to reproduce?

**Comment by Datseris** [comment]
```julia
using Makie

scene = lines(rand(100), rand(100), color = :red)
display(scene)

function set_limits!(scene, rect)
    AbstractPlotting.update_limits!(scene, rect)
    AbstractPlotting.scale_scene!(scene)
    AbstractPlotting.center!(scene)
    AbstractPlotting.update!(scene)
end

rect = Observable(FRect2D([0, 0], [1, 1]))
#
on(rect) do r
    set_limits!(scene, r)
end

# Then do:
rect[] = FRect2D([0.2, 0.2], [0.4, 0.4])
```

![image](https://user-images.githubusercontent.com/19669089/51317071-e996e200-1a56-11e9-8fd3-2c291ba461f6.png)

**Comment by SimonDanisch** [comment]
ah, yeah, that's because it still uses the boundingbox of the scene...
If this works for you:
```Julia


using Makie

function limit_center!(scene::Scene, padding = 0.01)
    bb = boundingbox(scene[Axis])
    bb = AbstractPlotting.transformationmatrix(scene)[] * bb
    w = widths(bb)
    padd = w .* padding
    bb = FRect3D(minimum(bb) .- padd, w .+ 2padd)
    AbstractPlotting.update_cam!(scene, bb)
    scene
end

scene = lines(rand(100), rand(100), color = :red)
display(scene)

function set_limits!(scene, rect)
    AbstractPlotting.update_limits!(scene, rect)
    AbstractPlotting.scale_scene!(scene)
    limit_center!(scene)
end

rect = Observable(FRect2D([0, 0], [1, 1]))
#
on(rect) do r
    set_limits!(scene, r)
end

# Then do:
rect[] = FRect2D([0.2, 0.2], [0.4, 0.4])
```
I should consider to update the definition of `center!` to work like this

**Comment by Datseris** [comment]
Yeah running the above seems to work. It doesn't crop though, it gives

![image](https://user-images.githubusercontent.com/19669089/51318165-d9343680-1a59-11e9-9d7e-aa1d14e8d779.png)

but cropping is another issue entirely.

Why don't we define 
```julia
function set_limits!(scene, rect, padding = 0.01)
    AbstractPlotting.update_limits!(scene, rect)
    AbstractPlotting.scale_scene!(scene)
    limit_center!(scene, padding)
end
```
as is, add a documentation string and export it? It is generally useful for many people.

Once this is exported I can open another issue to add cropping support, like
`set_limits(scene, rect, padding = 0.01; crop = true)` which also crops the data by default (which is what most people would want).

**Comment by asinghvi17** [comment]
Pinging @SimonDanisch - is this something we'd like to work on for the next release of AbstractPlotting?

**Comment by SimonDanisch** [comment]
Yeah, that's sort of implied by the limit cropping & better preserving of limits etc ;)
I'll add it explicitly!

**Comment by haberdashPI** [comment]
I think my problem is related, but I'm not sure. I just want a simple way to limit the range of the displayed data, when plotting multiple data sets. So if the data range is from 0 to 10 I want to be able to specify something like `ylim = (0,3)` and the resulting plot should behave as if the data range was 0 to 3 and crop the data outside the plot view. 

For example

```julia
using Makie
using Distributions
x1 = cumsum(rand(Normal(),100))
x2 = cumsum(rand(Normal(),100))
lim = FRect(-10,-10,100,4)
scene = Scene(limits=lim)
lines!(scene,axes(x1)[1],x1)
lines!(scene,axes(x2)[1],x2,color=:blue)
```

Results in the following plot:

![image](https://user-images.githubusercontent.com/7861592/57854195-98ed8680-77b5-11e9-99c2-91b37f28a4fd.png)
